rust/src/main.rs

```diff
@@ -1,35 +1,115 @@
+use std::collections::HashMap;
 use std::io;
 
+use rustemon::client::RustemonClient;
+use rustemon::model::pokemon::Pokemon;
+
+extern crate strum;
+#[macro_use]
+extern crate strum_macros;
+
+#[derive(EnumString)]
+enum Caught {
+    Y,
+    N,
+}
+
+#[derive(EnumString)]
+enum Funcs {
+    Encounter,
+    Withdraw,
+    Exit,
+}
+
 #[tokio::main]
 async fn main() {
-    let exit = String::from("exit\n");
     let rustemon_client: rustemon::client::RustemonClient =
         rustemon::client::RustemonClient::default();
+    let mut pokemon_box = HashMap::new();
 
     loop {
-        println!("Enter a species name (or `exit`):");
-        let mut input = String::new();
-
-        io::stdin()
-            .read_line(&mut input)
-            .expect("Failed to read line");
+        let mut function_input = String::new();
+        println!("Encounter, Withdraw or Exit?");
+        read_in(&mut function_input);
 
-        if input.eq_ignore_ascii_case(&exit) {
-            break;
+        match function_input.trim().parse() {
+            Ok(func) => match func {
+                Funcs::Encounter => encounter(&rustemon_client, &mut pokemon_box).await,
+                Funcs::Withdraw => withdraw(&pokemon_box),
+                Funcs::Exit => {
+                    break;
+                }
+            },
+            Err(err) => println!("Unrecognised command: {err}"),
         }
+    }
+}
+
+async fn encounter(client: &RustemonClient, pokemon_box: &mut HashMap<String, Pokemon>) {
+    let mut species_input = String::new();
+    println!("Enter a species name:");
+    read_in(&mut species_input);
+
+    let result = rustemon::pokemon::pokemon::get_by_name(&species_input, client).await;
+    match result {
+        Ok(species) => {
+            print_species(&species);
+
+            let mut caught_input = String::new();
+            println!("Did you catch it? (Y/N):");
+            read_in(&mut caught_input);
 
-        let result = rustemon::pokemon::pokemon::get_by_name(&input, &rustemon_client).await;
-        match result {
-            Ok(species) => {
-                for stat in species.stats.into_iter() {
-                    println!("{}: {}", stat.stat.name, stat.base_stat);
+            match caught_input.trim().parse() {
+                Ok(caught) => {
+                    match caught {
+                        Caught::Y => catch(pokemon_box, species),
+                        Caught::N => {
+                            println!("Better luck next time!")
+                        }
+                    };
```

Comment:
I would merge this match with the outer one (and the same for the function loop above
```suggestion
                Ok(Caught::Y) => catch(pokemon_box, species),
                Ok(Caught::N) => println!("Better luck next time!"),
                Err(err) => println!(
                    "Unable to fetch {} ({}): is the species name correct?",
                    species_input, err
                ),
```

You can also use the formatting values directly instead of passing them as extra args to `println!`

```rust
println!("Unable to fetch {species_input} ({err}): is the species name correct?")
```